### ISSUE TITLE
add callout capability to markers

### ISSUE THREAD

**Issue by incanus** [mapbox/DEPRECATED-mapbox-ios-sdk#23]
It would be nice to have something like `MKPinAnnotationView` with a `canShowCallout` property that automatically manages display and behavior of a callout similar to MapKit.

**Comment by incanus** [comment]
Note that this can probably be faked just like user location with a static view that is shown until the map moves.

**Comment by ivgalic** [comment]
It's possible to fake it, but it would be a nice feature :)

**Comment by incanus** [comment]
Well, by "fake" I mean only one callout is shown at a time, so it can be a view that's superimposed on the map at the point of the pin or whatever marker. When the map is tapped or moved, the callout disappears, so it doesn't have to travel with the map, which is more complicated and performance-intensive.

**Comment by ivgalic** [comment]
Well the biggest issue is that it would need to disappear as soon as the map is moved - we would want it to stay until the map is tapped. It's easy enough to calculate the screen position once a marker is tapped an add a view to the map at that point, but continuously tracking the position and moving the view as the map scrolls is both more complicated and could cause performance problems (I've just realized I've repeated your words :) ).
Another solution I was thinking of was to replace the pin's layer with another one that contains the callout, but I'm not sure if this will work (i.e. without flickering, re-animating the pin into place...).

**Comment by incanus** [comment]
I think this will solve it for us: http://nfarina.com/post/29883229869/callout-view

**Comment by spry** [comment]
Any ideas on how to add that to a map view and have it move?

**Comment by incanus** [comment]
I haven't checked out that API yet, but I'm thinking it could be done with the `mapViewRegionDidChange:` delegate method to update the attach point for the callout. When I work out this functionality natively, I'll likely follow the same pattern.

**Comment by schwanksta** [comment]
Just wanted to say I took @incanus' advice and used mapViewRegionDidChange with SMCalloutView and got it all working with map drag/zoom. The key was just to update the calloutView's frame origin.

I had to store the current annotation in a variable in the view controller on annotation tap in order to get this to work. This is so you can get the current location on the map to set the SMCalloutView to. Interestingly, I found you have to use the coordinate property of the annotation, because the point property doesn't get updated with each mapViewRegionDidChange event, but [mapView coordinateToPixel:currentAnnotation.coordinate] seems better in general. 

The +11 is specific to my app, because I have the offset moved up by 22px.

``` objective-c
- (void)mapViewRegionDidChange:(RMMapView *)mapView {
    if (calloutView.window) {        
        CGPoint pt = [mapView coordinateToPixel:currentAnnotation.coordinate];
        //objectAtIndex:3 is the bottomAnchor ImageView, aka the triangle.
        UIImageView *iv = [calloutView.subviews objectAtIndex:3];
        float widthadjust = iv.frame.size.width/2;
        float cx = iv.frame.origin.x + widthadjust;
        pt.x -= cx;
        pt.y -= iv.frame.size.height+11;
        calloutView.frame = (CGRect){ .origin=pt, .size=calloutView.frame.size };
    }

}
```

**Comment by alexiscreuzot** [comment]
I'd be very interested in a sample code for that specific matter. I'm still new to mapbox SDK and I'm struggling to achieve the same results that I would have with MKMapKit.

**Comment by incanus** [comment]
Assigned to [Tender discussion #239](http://support.mapbox.com/discussions/mapbox-ios-sdk/239-open-a-tooltip-on-rmmarker-touch).

**Comment by incanus** [comment]
@ivgalic @spry @kirualex just a ping that I've got some code in a branch (above) for this to start testing. I know it's got some issues when zoomed far out, as well as during compass heading rotation, but I'll be working on that.

**Comment by incanus** [comment]
The `23-callouts` branch is still progressing and basically done. You can see it in action in the [`callouts`](https://github.com/mapbox/weekend-picks-template-ios/tree/callouts) branch of the Weekend Picks sample app.

**Comment by incanus** [comment]
This just needs a slight amount of code cleanup in `RMMapView.m` to be ready, plus one minor fix from nfarina/calloutview#13.

**Comment by incanus** [comment]
This is merged into `develop` as of cff44eb9b7e2ddfc7889c60926e1efce7e6b13d7.

**Comment by incanus** [comment]
Best example to check this out is to consult the Weekend Picks sample project: 

https://github.com/mapbox/weekend-picks-template-ios
